ocikms: Oracle Cloud KMS support

deblasis opened the pull request:
Hi there, I found myself needing this so I thought it was a good opportunity to contribute a little.

Happy to make the necessary changes, I tried to comply with style as much as possible. 

Fixes #981 

cc: @colemickens, @davidpinhas, @b-dean

## Basic Demo

https://github.com/mozilla/sops/assets/29378614/f68a7a34-95a0-4a1c-846c-c61f1a091d5a

deblasis commented on Makefile:
```diff
@@ -36,7 +36,7 @@ generate: keyservice/keyservice.pb.go
 
 %.pb.go: %.proto
-	protoc --go_out=plugins=grpc:. $<
+	protoc --go_out=. --go-grpc_out=require_unimplemented_servers=false:.  $<
```
This should be the current way of generating files with protoc since 1.20 I believe.

The generated files are also split now as you can see below.

The flag `require_unimplemented_servers=false` is necessary to preserve the previous functionality.

Please note that I am using: 
![image](https://github.com/mozilla/sops/assets/29378614/64596de0-0293-4b5f-9429-7b12867d9f3c)

I have also tried 
```
protoc-gen-go v1.23.0
protoc        v3.13.0
```
but I am still getting errors with the plugin settings.

I don't think this is an issue but just flagging for the reviewer

deblasis commented on keyservice/keyservice.proto:
```diff
@@ -1,5 +1,7 @@
 syntax = "proto3";
 
+option go_package = "/keyservice";
```
Added this ad best practice and because I was getting errors because of this as well when doing `make generate`

deblasis committed:
feat: OCI KMS provider

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

deblasis committed:
docs: rstcheck fix

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

deblasis committed:
Merge branch 'main' of github.com:mozilla/sops into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'main' of github.com:mozilla/sops into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

jpark0910 commented:
Is there anything else blocking this PR from being merged main for the 3.10.0 milestone release? is there an ETA on 3.10.0?

felixfontein commented:
I personally won't have time to review this anytime soon (see https://github.com/getsops/sops/pull/1097#issuecomment-2607904325). I'm not sure if anyone else from the maintainers has time either, but it might well be that this won't make it into 3.10 and will get moved to the 3.11 milestone (once that exists).

masonhuemmer commented:
Any update?

deblasis committed:
Merge branch 'issue/#981-oci-kms' of github.com:deblasis/sops into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'master' into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
build(deps): Update oci-go-sdk to v65.101.0

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
feat(oci): Add OCI CLI environment provider and related tests

- Introduced `oci-cli-env-provider` for enhanced OCI configuration management.
- Implemented `configurationProvider` to prioritize CLI environment variables.
- Added tests for configuration provider functionality, ensuring correct behavior with OCI CLI and environment variables.

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs: Update README with OCI KMS authentication details and examples

- Clarified OCI KMS authentication order and added detailed examples for using OCI CLI and SDK environment variables.
- Improved formatting and removed unnecessary whitespace for better readability.

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge remote-tracking branch 'origin/master' into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(go): Update Go version to 1.24.0 and specify toolchain version

- Bumped Go version from 1.23.0 to 1.24.0.
- Added toolchain version 1.24.4 for consistency.

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge remote-tracking branch 'origin/master' into issue/#981-oci-kms

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

b-dean commented on ocikms/keysource.go:
```diff
@@ -0,0 +1,187 @@
+	}
+	region := parts[3]
+	vault_ref := parts[4]
+	return region, vault_ref, nil
```
@deblasis I'm not really sure what to do with this, because I assume the ocid in your demo video is real, but mine ends up with a different ocid and a different endpoint

My ocid is of this form:

```
ocid1.vault.oc1.iad.example.aaaaaREDACTED
```

which then my endpoint ends up being:
```
https://example-crypto.kms.us-ashburn-1.oraclecloud.com
```

with `example` there being what you call `vault_ref` here. 

So I'm not sure if we can extract the parts reliably from the key OCID. Having worked w/ OCI for a while this doesn't surprise me. They probably have several different iterations of vaults and the endpoints aren't consistent.

The best I can think of is either:
- also require the vault ocid and call [GetVault](https://docs.oracle.com/en-us/iaas/tools/go-sdk-examples/65.101.0/keymanagement/GetVault.go.html) to get the crypto endpoint
- use [ListVaults](https://docs.oracle.com/en-us/iaas/api/#/en/key/release/VaultSummary/ListVaults) to list all the vaults in the compartment, and then use the crypto management endpoint on each to call [ListKeys](https://pkg.go.dev/github.com/oracle/oci-go-sdk/v65@v65.101.0/keymanagement#KmsManagementClient.ListKeys) until you find the key that was specified by the sops args and then return the crypto endpoint url of the vault that has that key 🤮 

it'd be really nice to only have to say the key ocid, but that second option of loop through all of them seems horrible. any other ideas?

I suppose you could also have:
- try both versions of the crypto endpoint url (such as both `iad` and `us-ashburn-1`) with some sort of short timeout to figure out which is the correct endpoint.

b-dean replied:
@deblasis, good news! there's an easy way to deal with this. behold:

```go
package main

import (
        "fmt"
        "strings"

        "github.com/oracle/oci-go-sdk/v65/common"
)

func main() {
        keys := []string{
                "ocid1.key.oc1.uk-london-1.asdadsasdagz5aacmg.abwgiljtjasdasdasdagugpfe7wrtngukihgkybqxcoozz7sbh6lq",
                "ocid1.key.oc1.iad.asdadsasdagz5aacmg.abwgiljtjasdasdasdagugpfe7wrtngukihgkybqxcoozz7sbh6lq",
        }

        for i, ocid := range keys {
                endpoint, err := endpoint(ocid)
                if err != nil {
                        fmt.Println(err.Error())
                } else {
                        fmt.Printf("key %d endpoint: %s\n", i, endpoint)
                }
        }
}

func endpoint(ocid string) (string, error) {
        region, vaultExt, err := extractRefs(ocid)
        if err != nil {
                return "", err
        }

        cryptoEndpointTemplate := common.StringToRegion(region).EndpointForTemplate("kms", "https://{vaultExt}-crypto.kms.{region}.{secondLevelDomain}")
        cryptoEndpoint := strings.Replace(cryptoEndpointTemplate, "{vaultExt}", vaultExt, 1)
        return cryptoEndpoint, nil
}

const ocidParts = 6

func extractRefs(ocid string) (string, string, error) {
        parts := strings.Split(ocid, ".")
        if len(parts) != ocidParts {
                return "", "", fmt.Errorf("OCID length is %s, expected %d", ocid, ocidParts)
        }
        region := parts[3]
        vaultExt := parts[4]
        return region, vaultExt, nil
}
```

which outputs the following:
```
$ go run main.go
key 0 endpoint: https://asdadsasdagz5aacmg-crypto.kms.uk-london-1.oraclecloud.com
key 1 endpoint: https://asdadsasdagz5aacmg-crypto.kms.us-ashburn-1.oraclecloud.com
```

based on the your `keysource.go` and this [example I found in the oci-go-sdk](https://github.com/oracle/oci-go-sdk/blob/2d0ed3015dc7a5fa5d954d0d51c0fa95dc04f42a/example/container_image_signing/utils.go#L323-L332)

b-dean reviewed (changes requested):
going to need some way to provide the correct crypto endpoint (or find it)

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

b-dean committed:
use StringToRegion and EndpointForTemplate to create the crypto endpoint

Signed-off-by: Ben Dean <ben.dean@finvi.com>

deblasis committed:
Merge branch 'issue/#981-oci-kms' of github.com:deblasis/sops into issue/#981-oci-kms

deblasis committed:
feat: add early exit optimization for Instance Principal in configuration provider

- Implemented tests to verify that Instance Principal is skipped when valid environment variables are provided.
- Added fallback mechanism to ensure Instance Principal is called when environment variables are missing or invalid.
- Updated comments for clarity on the configuration provider's behavior regarding Instance Principals.

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
feat: implement lazy initialization for ConfigurationProvider

- Added lazyConfigurationProvider to defer the creation of the ConfigurationProvider until the first method call, optimizing performance for expensive providers like Instance Principal.
- Implemented tests to ensure factory is only called once and correctly propagates errors.
- Updated comments for clarity on the lazy initialization behavior.

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis commented:
> @deblasis , that makes sense. It'd be nice if Oracle would've just put some of this in their sdk to begin with. 

> I mainly published that because I wanted to use the `OCI_CLI_*` env and not have to make `~/.oci/config` everywhere I'd like to use sops.

I think the best thing you can do is to file a PR [upstream](https://github.com/oracle/oci-go-sdk/) to add native support for OCI_CLI_* env straight into the SDK! :)

deblasis committed:
Merge branch 'main' into issue/#981-oci-kms

deblasis commented:
Test failed due to a race condition originating from the SDK code itself.

I filed a PR over there with repro steps and a fix 

https://github.com/oracle/oci-go-sdk/pull/603

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/#981-oci-kms

# Conflicts:
#	README.rst
#	cmd/sops/main.go
#	config/config.go
#	go.mod
#	go.sum
#	keyservice/keyservice.go
#	keyservice/keyservice.pb.go
#	keyservice/keyservice.proto
#	keyservice/server.go
#	stores/stores.go

deblasis committed:
docs: add Oracle Cloud KMS to the supported providers list

ianmuge commented:
Is this still being considered?

harrisonquick commented:
Hi all,

Also curious if this is still being considered. We have a production use case for OCI KMS support; @deblasis I see you're still engaged - would help with rebasing / conflicts / testing / other updated be useful? Any chance a maintainer could confirm whether this PR is still a candidate for merging?

deblasis commented:
Hi @harrisonquick, I think the PR is blocked on https://github.com/getsops/sops/issues/2016 

As far as I understand it, and my memory might be a bit clouded (it has been a while), the maintainers want to move towards a pluggable system where KMSs can be added without making changes to SOPS itself, understandable.

Not sure what it is blocking that effort currently.

If you don't need to be bleeding edge, just use my fork and tweak it as you with until the pluggable system is ready.
You can always pull from upstream any change into it and use it. That's the beauty of OSS.

I had a production use case too and it worked very well.

harrisonquick commented:
@deblasis Thank you for your update! I will try to dig in further on the pluggable architecture status / project direction, and will happily use your fork until that becomes available.

deblasis commented:
> @deblasis Thank you for your update! I will try to dig in further on the pluggable architecture status / project direction, and will happily use your fork until that becomes available.

You are welcome @harrisonquick, you actually inspired me into taking action so thank you! 
I went on and built something that might give shape to the final solution or become it. Everyone wins either way. 

All the details are here: https://github.com/getsops/sops/issues/2016#issuecomment-5456092313
